core/src/main/java/jenkins/util/VirtualFile.java

```diff
@@ -518,15 +572,35 @@ public boolean supportIsDescendant() {
         @Override
         @Restricted(NoExternalUse.class)
         public boolean isDescendant(String potentialChildRelativePath) throws IOException {
+            if (potentialChildRelativePath.isEmpty() && cacheDescendant) {
+                return true;
+            }
+
             if (new File(potentialChildRelativePath).isAbsolute()) {
                 throw new IllegalArgumentException("Only a relative path is supported, the given path is absolute: " + potentialChildRelativePath);
             }
-            
+
+            // shortcut for direct child to avoid the complexity of the whole computation
+            // as we know that a file that is a direct descendant of its parent can only be descendant of the root
+            // if the parent is descendant AND the file itself is not symbolic
+            File directChild = new File(f, potentialChildRelativePath);
+            if (directChild.getParentFile().equals(f)) {
+                // potential shortcut for "simple" / direct child
+                if (!Util.isSymlink(directChild)) {
+                    return true;
+                }
+            }
+
             FilePath root = new FilePath(this.root);
             String relativePath = computeRelativePathToRoot();
             
             try {
-                return root.isDescendant(relativePath + potentialChildRelativePath);
+                boolean isDescendant = root.isDescendant(relativePath + potentialChildRelativePath);
+                if (isDescendant && potentialChildRelativePath.isEmpty()) {
+                    // in DirectoryBrowserSupport#zip, multiple calls to isDescendant are done for the same VirtualFile
+                    cacheDescendant = true;
+                }
+                return isDescendant;
             }
             catch (InterruptedException e) {
                 return false;
```

Comment:
can be handled properly by extending API of the restricted method

Reply:
@oleg-nenashev I do not see any reason to do that when the rest of the methods "just" wrap it within `IOException`, the behavior will not be consistent.